### PR TITLE
71168 71665 tag function model

### PR DESCRIPTION
**Pull request by janid1967**
TagFunction Model with migration
Endpoint getting TagFunction details. To be used in library when administrating Requirements on TagFunctions.
Endpoint getting TagFunctions which have any requirement. To be used in client wizard, page 1, when getting tags from Main for autoscoping. Should get all tags from main which have any TagFunction where TagFunction have any preservation requirement

### CODE REVIEW

File: src/Equinor.Procosys.Preservation.Domain/AggregateModels/TagFunctionAggregate/TagFunction.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            {
+                throw new ArgumentException($"Can't relate item in {requirement.Schema} to item in {Schema}");
+            }
+
+            _requirements.Add(requirement);
+        }
+
+        public void SetCreated(Person createdBy)
+        {
+            CreatedAtUtc = TimeService.UtcNow;
```
Comment (TordJoranger) [inline src/Equinor.Procosys.Preservation.Domain/AggregateModels/TagFunctionAggregate/TagFunction.cs]: nullcheck?

File: src/Equinor.Procosys.Preservation.Domain/AggregateModels/TagFunctionAggregate/TagFunction.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            CreatedById = createdBy.Id;
+        }
+
+        public void SetModified(Person modifiedBy)
+        {
+            ModifiedAtUtc = TimeService.UtcNow;
```
Comment (TordJoranger) [inline src/Equinor.Procosys.Preservation.Domain/AggregateModels/TagFunctionAggregate/TagFunction.cs]: nullcheck here to if we want to be consistent in how we handle setters in the class.

File: src/Equinor.Procosys.Preservation.Domain/AggregateModels/TagFunctionAggregate/TagFunction.cs (hunk 1 above)
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Domain/AggregateModels/TagFunctionAggregate/TagFunction.cs]: No need. TimeService is static

File: src/Equinor.Procosys.Preservation.Query/ResponsibleAggregate/GetAllResponsiblesQueryHandler.cs [hunk 3]
```diff
@@ -10,7 +10,7 @@ namespace Equinor.Procosys.Preservation.Query.ResponsibleAggregate
 {
     public class GetAllResponsiblesQueryHandler : IRequestHandler<GetAllResponsiblesQuery, Result<IEnumerable<ResponsibleDto>>>
     {
-        private readonly IResponsibleRepository _responsibleRepository;
+        private readonly IResponsibleRepository _responsibleRepository; //todo rewrite to use IReadOnlyContext
```
Comment (TordJoranger) [inline src/Equinor.Procosys.Preservation.Query/ResponsibleAggregate/GetAllResponsiblesQueryHandler.cs]: If forgotten?

File: src/tests/Equinor.Procosys.Preservation.Domain.Tests/AggregateModels/TagFunctionAggregate/TagFunctionRequirementTests.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        private TagFunctionRequirement _dut;
+        private Mock<RequirementDefinition> _requirementDefinitionMock;
+
+        [TestInitialize]
+        public void Setup()
+        {
+            _requirementDefinitionMock = new Mock<RequirementDefinition>();
+            _requirementDefinitionMock.SetupGet(r => r.Schema).Returns(TestPlant);
+            _requirementDefinitionMock.SetupGet(r => r.Id).Returns(5);
+            _dut = new TagFunctionRequirement(TestPlant, 4, _requirementDefinitionMock.Object);
```
Comment (TordJoranger) [inline src/tests/Equinor.Procosys.Preservation.Domain.Tests/AggregateModels/TagFunctionAggregate/TagFunctionRequirementTests.cs]: dut?

Comment (TordJoranger) [reply src/tests/Equinor.Procosys.Preservation.Domain.Tests/AggregateModels/TagFunctionAggregate/TagFunctionRequirementTests.cs]: device under test? ok figured it out eventually. (hate it :) )

File: src/Equinor.Procosys.Preservation.Domain/AggregateModels/TagFunctionAggregate/TagFunction.cs (hunk 1 above)
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Domain/AggregateModels/TagFunctionAggregate/TagFunction.cs]: done

File: src/Equinor.Procosys.Preservation.Query/ResponsibleAggregate/GetAllResponsiblesQueryHandler.cs (hunk 3 above)
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Query/ResponsibleAggregate/GetAllResponsiblesQueryHandler.cs]: No. Own BPI on it

File: src/tests/Equinor.Procosys.Preservation.Domain.Tests/AggregateModels/TagFunctionAggregate/TagFunctionRequirementTests.cs (hunk 4 above)
Comment (janid1967) [reply src/tests/Equinor.Procosys.Preservation.Domain.Tests/AggregateModels/TagFunctionAggregate/TagFunctionRequirementTests.cs]: **D**evice**U**nder**T**est.
 @hbjorgo thought me, and I adopted it

Comment (janid1967) [reply src/tests/Equinor.Procosys.Preservation.Domain.Tests/AggregateModels/TagFunctionAggregate/TagFunctionRequirementTests.cs]: shoot @hbjorgo

Comment (janid1967) [reply src/tests/Equinor.Procosys.Preservation.Domain.Tests/AggregateModels/TagFunctionAggregate/TagFunctionRequirementTests.cs]: > (hate it :) 
shoot @hbjorgo :)

File: src/tests/Equinor.Procosys.Preservation.Query.Tests/GetTagFunctionDetails/GetTagFunctionDetailsQueryHandlerTests.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                Assert.AreEqual(_tfWithRequirement.Description, tagFunction.Description);
+                Assert.AreEqual(_tfWithRequirement.RegisterCode, tagFunction.RegisterCode);
+                var tagFunctionRequirements = tagFunction.Requirements.ToList();
+                Assert.AreEqual(1, tagFunctionRequirements.Count);
+                Assert.AreEqual(_requirementDefinition.Id, tagFunctionRequirements.Single().RequirementDefinitionId);
+            }
+        }
+
+        [TestMethod]
+        public async Task HandleGetTagFunctionQuery_TagFunctionWithoutRequirement_ShouldReturnTagFunctionWithoutRequirement()
```
Comment (TordJoranger) [inline src/tests/Equinor.Procosys.Preservation.Query.Tests/GetTagFunctionDetails/GetTagFunctionDetailsQueryHandlerTests.cs]: I'm having trouble figuring this one out, We're using GetTagFunctionsHavingRequirementQueryHandler to fetch tagfunctions without requirements? 

Just double check that this is what is intended.

Comment (janid1967) [reply src/tests/Equinor.Procosys.Preservation.Query.Tests/GetTagFunctionDetails/GetTagFunctionDetailsQueryHandlerTests.cs]: Is the test unclear or the code itself?

Comment (janid1967) [reply src/tests/Equinor.Procosys.Preservation.Query.Tests/GetTagFunctionDetails/GetTagFunctionDetailsQueryHandlerTests.cs]: There is something fishy here .. I fix

Comment (janid1967) [reply src/tests/Equinor.Procosys.Preservation.Query.Tests/GetTagFunctionDetails/GetTagFunctionDetailsQueryHandlerTests.cs]: The class inside file GetTagFunctionDetails/GetTagFunctionDetailsQueryHandler.cs had wrong name (rename issue)

**Pull request merged**